File: Data Structures & Algorithms/non-overlapping-intervals/submission-16.js
class Solution {
    /**
     * @param {number[][]} intervals
     * @return {number}
     */
    eraseOverlapIntervals(intervals) {
        intervals.sort((a, b) => a[1] - b[1]);
        const n = intervals.length;
        let memo = new Array(n).fill(-1);  

        const dfs = (i) => {
            if (i >= n) return 0;
            if (memo[i] !== -1) return memo[i];

            let res = 1;
            for (let j = i + 1; j < n; j++) {
                if (intervals[i][1] <= intervals[j][0]) {
                    res = Math.max(res, 1 + dfs(j));
                }
            }
            memo[i] = res;
            return res;
        };

        const maxNonOverlapping = dfs(0);
        return n - maxNonOverlapping;
    }
}

File: Data Structures & Algorithms/count-subsequences/submission-16.cpp
class Solution {
public:
    int numDistinct(string s, string t) {
        int m = s.size(), n = t.size();
        if (n > m) return 0;
        vector<vector<int>> dp(m + 1, vector<int>(n + 1, -1));
        return dfs(s, t, 0, 0, dp);
    }

private:
    int dfs(const string &s, const string &t, int i, int j, vector<vector<int>> &dp) {
        if (j == t.size()) return 1;
        if (i == s.size()) return 0;
        if (dp[i][j] != -1) return dp[i][j];

        int res = dfs(s, t, i + 1, j, dp);
        if (s[i] == t[j]) {
            res += dfs(s, t, i + 1, j + 1, dp);
        }
        dp[i][j] = res;
        return res;
    }
};